cmake/IBAMRTesting.cmake

```diff
@@ -0,0 +1,293 @@
+    return ()
+  endif ()
+
+  # Add test source if it exists (may not because of _2d/_3d suffix)
+  if (EXISTS "${IBAMR_TEST_DIR}/${test_DIRECTORY}/${name}.cpp")
+    list(APPEND test_SOURCES "${IBAMR_TEST_DIR}/${test_DIRECTORY}/${name}.cpp")
+  endif ()
+
+  # If there aren't any sources, the this is not going to work...
+  if (NOT test_SOURCES)
+    message(FATAL_ERROR "No sources detected for test: ${test_DIRECTORY}/${name}")
+  endif ()
+
+  # Find all of the test inputs
+  file(GLOB input_files CONFIGURE_DEPENDS "${IBAMR_TEST_DIR}/${test_DIRECTORY}/${name}.*.input")
+  if (NOT input_files)
+    return ()
+  endif ()
+
+  ###############################################################################################
+  # Create build targets
+
+  # Add a top level test build target
+  if (NOT TARGET "ibamr-tests")
+    add_custom_target(ibamr-tests ALL)
+  endif ()
+
+  # Add a directory build target and make the top level depend on it
+  if (NOT TARGET "ibamr-tests-${test_DIRECTORY}")
+    add_custom_target("ibamr-tests-${test_DIRECTORY}")
+    add_dependencies(ibamr-tests "ibamr-tests-${test_DIRECTORY}")
+  endif ()
+
+  # Create the test executable
+  set(_target "ibamr-test-${test_DIRECTORY}-${name}")
+  add_executable(${_target} EXCLUDE_FROM_ALL "${test_SOURCES}")
+  set_target_properties(${_target}
+    PROPERTIES
+    RUNTIME_OUTPUT_DIRECTORY
+      "${CMAKE_BINARY_DIR}/tests/${_dir}"
+    OUTPUT_NAME
+      ${name}
+    )
+  target_link_libraries(${_target} PRIVATE ${test_LIBRARIES})
+  target_compile_definitions(${_target}
+    PRIVATE
+      SOURCE_DIR="${IBAMR_TEST_DIR}/${test_DIRECTORY}/")
+  # Connect the target to the directory target
+  add_dependencies("ibamr-tests-${test_DIRECTORY}" ${_target})
+
+  # Add the tests for this target
+  foreach (input ${input_files})
+    get_filename_component(input_stripped ${input} NAME_WLE)
+    get_filename_component(input_directory ${input} DIRECTORY)
+    set(check_output_file "${input_directory}/${input_stripped}.output")
+
+    _ibamr_parse_test(${input_stripped})
+
+    set(case "${test_DIRECTORY}-${input_stripped}")
+    set(test_workdir "${CMAKE_BINARY_DIR}/Testing/${case}")
+    set(test_labels ${test_DIRECTORY})
+    set(test_list)
+
+
+    set(_launcher)
+    set(_extra_args)
+    set(load 1)
+    # Add MPI flags
+    if (n_mpirun GREATER 1)
+      set(load ${n_mpirun})
+      # Oversubscribe
+      set(_mpi_flags ${MPIEXEC_PREFLAGS} --bind-to none)
+      if (n_mpirun GREATER IBAMR_MAX_PARALLEL)
+        if (IBAMR_ALLOW_OVERSUBSCRIBED)
+          message(WARNING "Adding a test that is oversubscribed")
+          list(APPEND _mpi_flags --oversubscribe)
+          set(load ${IBAMR_MAX_PARALLEL})
+        else ()
+          # Don't add tests that would require oversubscription
+          continue ()
+        endif ()
+      endif ()
+      list(APPEND _launcher
+        ${MPIEXEC} ${_mpi_flags} ${MPIEXEC_NUMPROC_FLAG} ${n_mpirun})
+      list(APPEND test_labels "MPI")
```

Comment:
I'm not sure where it should be added but we also need to disable multithreading with something like
```
export OMP_NUM_THREADS=1
export OMP_THREAD_LIMIT=1
```

Reply:
Okay, that will be added down below where I setup properties.